source/common/common/compiled_string_map.h

```diff
@@ -0,0 +1,235 @@
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <string>
+#include <vector>
+
+#include "envoy/common/pure.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+
+/**
+ * This is a specialized structure intended for static header maps, but
+ * there may be other use cases.
+ * The structure is:
+ * 1. a length-based lookup table so only keys the same length as the
+ * target key are considered.
+ * 2. a trie that branches on the "most divisions" position of the key.
+ *
+ * Using this structure is not recommended unless the table is
+ * initialize-once, use-many, as the "compile" operation is expensive.
+ *
+ * Unlike a regular trie, this structure cannot be used for prefix-based
+ * matching.
+ *
+ * For example, if we consider the case where the set of headers is
+ * `x-prefix-banana`
+ * `x-prefix-babana`
+ * `x-prefix-apple`
+ * `x-prefix-pineapple`
+ * `x-prefix-barana`
+ * `x-prefix-banaka`
+ *
+ * A standard front-first trie looking for `x-prefix-banana` would walk
+ * 7 nodes through the tree, first for `x`, then for `-`, etc.
+ *
+ * This structure first jumps to matching length, eliminating in this
+ * example case apple and pineapple.
+ * Then the "best split" node is on
+ *   `x-prefix-banana`
+ *               ^
+ * so the first node has 3 non-miss branches, n, b and r for that position.
+ * Down that n branch, the "best split" is on
+ *   `x-prefix-banana`
+ *                 ^
+ * which has two branches, n or k.
```

Comment:
IIUC the "best split" can be anywhere in the string. I think modifying the example to the following, will clarify how this should work.
```
 * `x-prefix-banana`
 * `x-prefix-banara`
 * `x-prefix-apple`
 * `x-prefix-pineapple`
 * `x-prefix-barana`
 * `x-prefix-banaka`
```
then the best split would've been in the character before-last.

Reply:
It can't be *anywhere* in the string, only at an index where there's a difference. The example is specifically showing that 3 branches would be "best" over 2 branches. I don't understand what your example clarifies that the existing example doesn't.

Reply:
My comment was looking at it from a "trie" perspective.
When I first read the comment, I thought that in each branch, the structure only keeps the "suffix" (after the branch point). I think that updating the example to a case where the branching happens due to a "best-split" that is not the first split, will reduce that confusion to future readers.

Reply:
Hopefully the md file explanation covers this sufficiently (though it doesn't do the second split earlier in the string).

If you'd like me to rearrange the strings in the new example so that it does the "last" index split first and the earlier one second I can do that, but I think the diagram having numbers in it *probably* makes it clear enough. (And late first would be just as potentially confusing as early first, and having three sequential splits would be getting so long as to be harder to read...)